macros/src/syntax/parse.rs

```diff
@@ -0,0 +1,496 @@
+                assert!(resources.is_none(), "duplicated resources field");
+
+                let tt = tts.next();
+                if let Some(TokenTree::Delimited(array)) = tt {
+                    assert_eq!(
+                        array.delim,
+                        DelimToken::Bracket,
+                        "expected bracket, found {:?}",
+                        array.delim
+                    );
+
+                    resources = Some(super::parse::idents(array.tts));
+
+                } else {
+                    panic!("expected array, found {:?}", tt);
+                }
+            }
+            id => panic!("unexpected field {}", id),
+        }
+
+        let tt = tts.next();
+        assert_eq!(
+            tt,
+            Some(TokenTree::Token(Token::Comma)),
+            "expected comma, found {:?}",
+            tt
+        );
+    }
+
+    (
+        local,
+        path.expect("path field is missing"),
+        resources.unwrap_or(HashSet::new()),
+    )
+}
+
+pub fn idle(tts: Vec<TokenTree>) -> Idle {
+    let (locals, path, resources) = idle_init(tts, true);
+
+    Idle {
+        local: locals.expect("local field is missing"),
+        path,
+        resources,
+    }
+}
+
+pub fn init(tts: Vec<TokenTree>) -> Init {
+    let (_, path, resources) = idle_init(tts, false);
+
+    Init { path, resources }
+}
+
+fn idents(tts: Vec<TokenTree>) -> HashSet<Ident> {
+    let mut idents = HashSet::new();
+
+    let mut tts = tts.into_iter();
+    while let Some(tt) = tts.next() {
+        if let TokenTree::Token(Token::Ident(id)) = tt {
+            assert!(!idents.contains(&id), "ident {} already listed", id);
+            idents.insert(id);
+
+            if let Some(tt) = tts.next() {
+                assert_eq!(tt, TokenTree::Token(Token::Comma));
+            } else {
+                break;
+            }
+        } else {
+            panic!("expected ident, found {:?}", tt);
+        };
+    }
+
+    idents
+}
+
+pub fn resources(tts: Vec<TokenTree>) -> Resources {
+    let mut resources = HashMap::new();
+
+    let mut tts = tts.into_iter();
+    while let Some(tt) = tts.next() {
+        let name = if let TokenTree::Token(Token::Ident(ident)) = tt {
+            ident
+        } else {
+            panic!("expected ident, found {:?}", tt);
+        };
+
+        assert!(
+            !resources.contains_key(&name),
+            "resource {} already listed",
+            name
+        );
+
+        let tt = tts.next();
+        assert_eq!(
+            tt,
+            Some(TokenTree::Token(Token::Colon)),
+            "expected comma, found {:?}",
+            tt
+        );
+
+        let mut pieces = vec![];
+        loop {
+            if let Some(tt) = tts.next() {
+                if tt == TokenTree::Token(Token::Eq) {
+                    break;
+                } else {
+                    pieces.push(tt);
+                }
+            } else {
+                panic!("expected type, found EOM");
```

Comment:
Nit: maybe expand `EOM` to `end of macro`? It took me a few moments to recognize.